RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc

```diff
@@ -0,0 +1,474 @@
+  clusterOnlySameBXRecHits  = ps.getParameter<bool>("clusterOnlySameBXRecHits");
+  // useGE21Short              = ps.getParameter<bool>("useGE21Short"); --> needed in GEMSegmentBuilder.cc
+
+  // maybe to be used in the future ???
+  // Pruning                = ps.getParameter<bool>("Pruning");
+  // BrutePruning           = ps.getParameter<bool>("BrutePruning");
+  // maxRecHitsInCluster is the maximal number of hits in a precluster that is being processed
+  // This cut is intended to remove messy events. Currently nothing is returned if there are
+  // more that maxRecHitsInCluster hits. It could be useful to return an estimate of the 
+  // cluster position, which is available.
+  // maxRecHitsInCluster    = ps.getParameter<int>("maxRecHitsInCluster");
+  // onlyBestSegment        = ps.getParameter<bool>("onlyBestSegment");
+
+  // CSC uses pruning to remove clearly bad hits, using as much info from the rechits as possible: charge, position, timing, ...
+  // In fits with bad chi^2 they look for the worst hit (hit with abnormally large residual) 
+  // if worst hit was found, refit without worst hit and select if considerably better than original fit.
+}
+
+/* Destructor
+ *
+ */
+GEMSegmentAlgorithm::~GEMSegmentAlgorithm() {
+}
+
+
+std::vector<GEMSegment> GEMSegmentAlgorithm::run(const GEMEnsemble& ensemble, const EnsembleHitContainer& rechits) {
+
+  theEnsemble = ensemble;
+
+  GEMDetId chId((theEnsemble.first)->id());
+  edm::LogVerbatim("GEMSegmentAlgorithm") << "[GEMSegmentAlgorithm::run] build segments in chamber " << chId;
+  
+  // pre-cluster rechits and loop over all sub clusters separately
+  std::vector<GEMSegment>          segments_temp;
+  std::vector<GEMSegment>          segments;
+  ProtoSegments rechits_clusters; // this is a collection of groups of rechits
+  
+  if(preClustering) {
+    // run a pre-clusterer on the given rechits to split obviously separated segment seeds:
+    if(preClustering_useChaining){
+      // it uses X,Y,Z information; there are no configurable parameters used;
+      // the X, Y, Z "cuts" are just (much) wider than reasonable high pt segments
+      edm::LogVerbatim("GEMSegmentAlgorithm") << "[GEMSegmentAlgorithm::run] preClustering :: use Chaining";
+      rechits_clusters = this->chainHits( rechits );
+    }
+    else{
+      // it uses X,Y information + configurable parameters
+      edm::LogVerbatim("GEMSegmentAlgorithm") << "[GEMSegmentAlgorithm::run] Clustering";
+      rechits_clusters = this->clusterHits(rechits );
+    }
+    // loop over the found clusters:
+      edm::LogVerbatim("GEMSegmentAlgorithm") << "[GEMSegmentAlgorithm::run] Loop over clusters and build segments";
+    for(auto sub_rechits = rechits_clusters.begin(); sub_rechits !=  rechits_clusters.end(); ++sub_rechits ) {
+      // clear the buffer for the subset of segments:
+      segments_temp.clear();
+      // build the subset of segments:
+      segments_temp = this->buildSegments( (*sub_rechits) );
```

Comment:
this would have 3x less allocations if buildSegments was instead

void buildSegments(const EnsembleHitContainer& rechits, std::vector<GEMSegment> &gemsegs);